### ISSUE TITLE
[Fix] AttentionMapLoss gets fed post-softmax

### ISSUE THREAD

**Issue by elouanzer** [elouanzer/shrinkai#11]
## Motivation

Two distinct problems, both stemming from how `CausalLMWrapper` (tutorial 03 `docs/tutorials/03_distillation_llm_generation.ipynb`, cell 4)
extracts attention maps from HuggingFace models, and both verified directly against the installed `transformers==5.16.1` source (not assumed):

```python
outputs = self.hf_model(
    input_ids=input_ids, attention_mask=attention_mask, output_attentions=True
)
features_dict[alias] = outputs.attentions[layer_idx]
```

## Problem 1: double softmax when `loss_type="kl"`

`AttentionMapLoss`'s `"kl"` branch (`src/shrinkai/distillation/losses/features.py`) runs `log_softmax(s_map)` / `softmax(t_map)` internally — by design, it expects *pre-softmax* attention scores (same convention as `"mse"`, per TinyBERT/MiniLM). `outputs.attentions` from any HuggingFace model is documented and has always been the *post*-softmax attention weights. So with the only attention-extraction wiring this library ships, choosing `loss_type="kl"` softmaxes the same map twice.

Confirmed this is genuinely unreachable any other way: GPT-2's `eager_attention_forward` (`transformers/models/gpt2/modeling_gpt2.py`) computes the raw scores and overwrites them with the softmaxed version in the same local scope, before returning — there's no submodule boundary a forward hook could attach to, and `output_attentions` only ever exposes the post-softmax tensor:

```python
attn_weights = torch.matmul(query, key.transpose(-1, -2)) * scaling   # raw scores — never exposed
attn_weights = nn.functional.softmax(attn_weights, dim=-1)            # overwritten here
attn_weights = nn.functional.dropout(attn_weights, p=dropout, training=module.training)
return attn_output, attn_weights                                       # what output_attentions gives you
```

For an already-normalized row ($p_i \in [0,1]$, $\sum_i p_i = 1$), a second softmax gives $e^{p_i} / \sum_j e^{p_j}$. Since $e^{p_i} \in [1, e]$ regardless of how peaked the original distribution was, the result is crushed toward uniform:

$$
\max_i \big(\text{softmax}(p)\big)_i \;\le\; \frac{e}{e + (N-1)}
$$

for a row of length $N$. With a causal mask, most rows have very few valid (non-zero) positions — at query position 0 there is exactly **one** valid key, so $N-1 = 127$ structurally-masked ("impossible") positions each come back at weight $e^0 = 1$, on equal footing with the one real position's $\le e$. That row's real position gets at most $e/(e+127) \approx 2.1\%$ of the mass — this bound is exact arithmetic, not measured. A synthetic test with realistically peaked causal attention rows (not GPT-2's actual activations) put the average spurious mass on structurally-impossible positions around 49% overall and ~98% at query position 0 specifically, consistent with that bound.

### Proposed fix

Per the "unreachable pre-softmax scores" point above, re-deriving raw scores isn't realistically on the table. The smaller, correct change: let `AttentionMapLoss` accept **probabilities** directly instead of assuming raw scores — e.g. an `input_format: Literal["logits", "probs"] = "logits"` constructor argument, where `"probs"` skips the internal softmax/log_softmax and does `torch.log(x.clamp(min=eps))` instead. Keeps the default behavior (and TinyBERT/MiniLM's original convention) unchanged for anyone feeding raw scores from a custom attention module; makes it correct for anyone stuck with `output_attentions`-style extraction, which is realistically most users going through HuggingFace.

## Problem 2: attention dropout leaks into the compared maps during training

Same `eager_attention_forward` snippet above: dropout is applied to the attention weights *before* they're returned, gated on `module.training`. `DistillationEngine` (`src/shrinkai/distillation/engine.py`) keeps `teacher.eval()` throughout and calls `student.train()` at the start of every epoch — confirmed by reading the engine directly. `distilgpt2`'s `attn_pdrop` is `0.1` (checked via `AutoConfig.from_pretrained("distilgpt2")`).

So during every training step, the student's extracted attention maps have ~10% of entries zeroed and the remainder scaled by $1/0.9$ (standard inverted dropout), compared against a teacher whose maps are always clean. This affects **both** `loss_type` options — it's not specific to `"kl"`, and it's live in tutorial 03 *today* (which uses `"mse"`), not just a latent risk for `"kl"`.

### Proposed fix — needs a decision, not just a patch

This is less clean than Problem 1 because attention dropout is doing real regularization work during training, not just getting in the way:

1. **Accept the noise, document it** — cheapest, but the comparison stays measurably biased in a way most users won't expect from a "feature distillation loss."
2. **Force the relevant attention submodules to eval-mode dropout while leaving the rest of the student in train-mode** — technically possible (`nn.Module.training` is a per-instance flag), removes the noise, but is surgical/fragile to implement generically across architectures, and silently changes the model's effective training regularization compared to training it without distillation.
3. **Document it as a known limitation of hidden-state/attention distillation in general** and leave it as a caveat rather than a "fix" — this tension (compare clean signals vs. keep realistic training dynamics) shows up for any dropout-containing intermediate layer used in feature distillation, not just attention maps.

## Open questions

- Is `input_format` the right shape for Problem 1's fix, or should `"kl"` and `"mse"` be split into genuinely separate classes/functions instead of branches on one class, given they now have diverging input conventions in practice?
- For Problem 2, is there an existing convention elsewhere in the library (or literature) for how much "cleanliness" feature-distillation losses should assume from a `.train()`-mode student, before deciding between options 1–3 above?
- Tutorial 03 needs updating either way once a fix direction is picked — its numbers (and `feature_weight=10.0`, already flagged in #10) were tuned against the current, affected behavior.

## References

- HuggingFace `output_attentions` semantics (post-softmax weights): standard, stable across `transformers` versions.
- `transformers/models/gpt2/modeling_gpt2.py::eager_attention_forward` — readdirectly from the installed `transformers==5.16.1` package for this issue.
